Add Javascript Coding Standard #91

sihaoo opened the pull request:
Fixes #91
Based on the HTML coding standards md from #93

damithc commented:
Looks like this got left out of the review.
1. Look at the Java/C++ standard .md and .html file and try to follow a similar style where possible.
2. Modify the link from README.md

sihaoo commented:
I have referenced mainly from the HTML standard file in #93 and copied the content from the link referenced in README.md. Do I recreate a new one to reference from Java/C++ instead?

damithc commented:
See #97 for an example.

wkurniawan07 commented on contents/coding-standards-javascript.md:
````diff
@@ -0,0 +1,155 @@
+
+    // G_O_O_D
+    var FEEDBACK_RESPONSE_RECIPIENT = 'responserecipient';
+    ```
````
Can you use another examples here? These examples are taken from TEAMMATES and may be confusing to developers not in that project.

wkurniawan07 commented on contents/coding-standards-javascript.md:
````diff
@@ -0,0 +1,155 @@
+    var FEEDBACK_RESPONSE_RECIPIENT = 'responserecipient';
+    ```
+
+4. Line wrapping rules for improving readability [follow TEAMMATES Java Style Guide files rule 3](https://docs.google.com/document/pub?id=1iAESIXM0zSxEa5OY7dFURam_SgLiSMhPQtU0drQagrs&embedded=true):
````
Point this link to the Java style guide in this repo.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,155 @@
+
+4. Line wrapping rules for improving readability [follow TEAMMATES Java Style Guide files rule 3](https://docs.google.com/document/pub?id=1iAESIXM0zSxEa5OY7dFURam_SgLiSMhPQtU0drQagrs&embedded=true):
+
+  When wrapping lines, the main objective is to improve readability. Feel free to break rules if it improves readability. Do not accept the auto-formatting suggested by the IDE as Eclipse’s JavaScript auto-formatting is very different from that suggested by TEAMMATES.
```
Again, don't mention TEAMMATES here.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,155 @@
+  mySum = a + b + c
+          + d;
+  myString = ‘Long line split’
+             + ‘split into two parts’;
```
Use the standard `'` instead of multibyte quotes so that GitHub can parse and highlight the syntax correctly.

wkurniawan07 commented on contents/coding-standards-javascript.md:
````diff
@@ -0,0 +1,155 @@
+        : gamma;
+  ```
+
+Main rule of the thumb: To be consistent with Java coding style in TEAMMATES
````
And here too.

sihaoo commented:
I will make the changes!

sihaoo commented:
Sorry for the late update! Was away.

1. Updated format to adhere to example #97
2. Modified the content based on comments given.
3. Included html file to reference to markdown file

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,164 @@
+ - __Indentation__
+   
+   Indentation should be 4 spaces for javascript files.
+   4 spaces should be used instead of 2 `spaces` or `tab`.
```
You don't need to wrap `spaces` as code.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,164 @@
+	
+ - __Line wrapping rules__ 
+ 
+  Line wrapping rules for improving readability <a href="coding-standards-java.html">[follow Java Style Guide files rule 3]</a>
```
This link won't work. Use [this](https://cdn.rawgit.com/nus-cs2103/website/master/contents/coding-standards-java.html).

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,164 @@
+   myFunction(arg1,
+             myOtherFunction(arg2,
+                             arg3),
+             arg4);
```
Indent so that it shows like this:
```javascript
myFunction(arg1,
           myOtherFunction(arg2,
                           arg3),
           arg4);
```

wkurniawan07 commented:
Also:

1. Modify the presentation of good/bad examples to follow the way used in #97.
2. Update the link to the coding style in `README.md`. [This](https://cdn.rawgit.com/nus-cs2103/website/master/contents/coding-standards-javascript.html) will be the link. It won't work if you try now but it will later on when this PR is merged.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,164 @@
+3. [Future of Javascript (ES6 - Incoming Features)](https://github.com/lukehoban/es6features)
+4. [ES6 Compatibility in Browsers](http://kangax.github.io/compat-table/es6/)
+5. [Useful CSS Selectors to know](http://code.tutsplus.com/tutorials/the-30-css-selectors-you-must-memorize--net-16048)
+6. [W3schools CSS Selectors Reference](http://www.w3schools.com/cssref/css_selectors.asp)
```
I don't think the last two references are relevant.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+			<pre lang="javascript">
+   // bad
+   var foobardescription = 'helloworld';
+			</pre>
```
This one is duplication of the first bad example.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+   var alphabets = [
+           a,
+           b,
+           c
```
Indent 4 spaces as mentioned in this guide.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+   var alphabets = [
+               a
+           ,   b
+           ,   c
```
Here too. The issue here is misplaced comma, not bad indentation.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+         one: 1,
+         two: 2,
+         three: 3,
+         four: 4
```
Here as well.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+               one: 1
+           ,   two: 2
+           ,   three: 3
+           ,   four: 4
```
And here.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+       myFunction(arg1,
+                   myOtherFunction(arg2,
+                                  arg3),
+                   arg4);
```
The alignment is wrong here. What is intended:
```javascript
myFunction(arg1, arg2
           arg3, arg4);
myFunction(arg1,
           myOtherFunction(arg2,
                           arg3),
           arg4);
```

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+  mySum = a  b  c
+               d;
+  myString = 'Long line split'
+                  'split into two parts';
```
I think you somehow removed all the `+`s.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+                                       : gamma;
+  alpha = (aLongBooleanExpression)
+            ? beta
+            : gamma;
```
```javascript
alpha = (aLongBooleanExpression) ? beta
                                 : gamma;
alpha = (aLongBooleanExpression)
      ? beta
      : gamma;
```

wkurniawan07 commented:
Some general comments:

1. You're still using tabs in some places. Change them to 2 or 4 spaces as appropriate.
2. You don't need the `// good` or `// bad` comments anymore as the table header will already tell.
3. Don't line break before closing `</pre>` because the line break will be rendered.

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+			</pre>
+		</td>
+	 </tr>
+   </table>
```
Actually, for this table, rather than making the good and the bad side-by-side (not balanced since there are 3 bad examples), you can make the table entirely sequential, for e.g the example of good/bad attributes in the [HTML coding standard](https://github.com/nus-cs2103/website/blob/master/contents/coding-standards-html.md).

wkurniawan07 commented on contents/coding-standards-javascript.md:
```diff
@@ -0,0 +1,273 @@
+             .highlight()
+             .end()
+        .find('.open')
+             .updateCount();
```
From `$('#responses')` to `.find('.selected')` should be 4 spaces?

wkurniawan07 commented:
@sihaoo I haven't checked on this for a while. Any updates?

damithc commented:
Closing due to inactivity. Reopen if you decide to work on this again @sihaoo
